Port tf2_kdl

vmayoral opened the pull request.

rhaschke committed:
improve computation efficiency

rhaschke committed:
improve comments

rhaschke committed:
normalize quaternions to be in half-space w >= 0

... as in tf1

rhaschke committed:
add Eigen::Isometry3d conversions

Fix conflicts and adapt code to ROS 2.0
Edited by @vmayoral.

vmayoral committed:
Port tf2_kdl

vmayoral committed:
Adapt for eigen3

tfoote commented on tf2_kdl/package.xml:
```diff
@@ -12,13 +12,13 @@
   <build_depend>tf2_ros</build_depend>
 
   <run_depend>cmake_modules</run_depend> <!-- needed for eigen -->
-  <run_depend>eigen</run_depend>
```
Eigen is Eigen3 by default. 
https://index.ros.org/d/eigen/
If you want eigen2 you call it out explicitly
https://index.ros.org/d/eigen2/

But I think we've effectively transtioned to eigen3 fully now based on the rosindex lack of usage of the older key.

vmayoral replied:
Understood. I recall having had some issues with that in OS X but can't reproduce it right now so I'll return to `eigen` in the package.xml.

tfoote reviewed (commented):
This would be great to have. Before merging this I'd like to see the unit tests ported too. They use a launch file in ROS1 but in ROS2 there's no master so that's not necessary for unit tests so everything can just be a unit test. The launch file only launches a single executable at a time anyway.

vmayoral committed:
Return to eigen in package.xml

Follow advice of @tfoote

vmayoral committed:
Remove cmake_modules

anasarrak committed:
Update tf2_kdl messages

vmayoral committed:
Include the right headers

vmayoral committed:
tf2_eigen, leftover from the cherry-pick

While cherry-picking changes to get isometry3d in

vmayoral committed:
Fix problem with stamp

Picked also a leftover from the ROS2 conversion

vmayoral committed:
Fix more issues while porting to ROS2

vmayoral committed:
Remove duplicated definitions

vmayoral committed:
Undo changes in unit tests

Leave code as it was left by @clalancette.
Only added some additional lines to test also
Isometry as per the commits of Robert

vmayoral committed:
Update tf2_eigen, add toMsg2

Convert a Eigen::Vector3d type to a geometry_msgs::msg::Vector3
while avoiding overloading issues with previous definitions

Commit:
make tf2_kdl an ament pkg

mlautman commented:
This port is still incomplete. Here are some issues that remain:

- [x] Finish package.xml port https://github.com/AcutronicRobotics/geometry2/pull/3
- [x] Port underlying [python_orcos_kdl](https://github.com/orocos/orocos_kinematics_dynamics/tree/master/python_orocos_kdl) for PyKDL 
- [ ] Remove roslib and rospy from [tf2_kdl](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/src/tf2_kdl/tf2_kdl.py)
- [ ] Remove deprecated [tf2_kdl/include/tf2_kdl/tf2_kdl subdirectory](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/include/tf2_kdl/tf2_kdl/tf2_kdl.h)
- [ ] Port the [CPP test](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/test/test_tf2_kdl.cpp) and python [test](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/scripts/test.py). Also, test.py should be moved to the test/ subdirectory instead of scripts and should not be built as an [executable](https://github.com/AcutronicRobotics/geometry2/blob/1020e530449ed2751d3cf7e52963ea9e6de1ee7b/tf2_kdl/CMakeLists.txt#L27) but as a python test.

mlautman commented on tf2_kdl/CMakeLists.txt:
```diff
@@ -1,44 +1,54 @@
-  DEPENDS Eigen orocos_kdl
-)
+if(NOT WIN32)
+  add_definitions(-std=c++11)
```
Why c++11? Why not 14. Add a note explaining this so that future devs know when it can be removed

vmayoral replied:
Thanks @mlautman 28d444e12eab3cc0c1be879426fbfa8f8cca48b2

mlautman commented on tf2_kdl/CMakeLists.txt:
```diff
@@ -1,44 +1,54 @@
+find_package(tf2_msgs REQUIRED)
+find_package(Eigen3 REQUIRED)
 
-catkin_python_setup()
```
python libs need to be exported with  
```cmake 
ament_python_install_package(${PROJECT_NAME}
    PACKAGE_DIR src/${PROJECT_NAME})
```

mlautman commented on tf2_kdl/CMakeLists.txt:
```diff
@@ -1,44 +1,54 @@
-endif()
-
-endif()
+# if(BUILD_TESTING)
```
Don't just comment out tests. This port still needs work which gets hidden because the tests are disabled.

vmayoral committed:
Default to C++14

vmayoral committed:
First steps towards getting tests

vmayoral committed:
Fix timePointFromTime conversions

vmayoral committed:
Make unit test compile

jacobperron commented:
@vmayoral I'm moved this back to "in progress", let me know if/when it's ready for another round of review.

vmayoral commented:
Sure @jacobperron, will do.

ahcorde committed:
Fixing test tf2_kdl

ahcorde committed:
including it own includes in tf2_kdl

vmayoral commented:
@ahcorde cherry-picked your commits here. Please review.

vmayoral commented:
@jacobperron, after @ahcorde's review, it should be ready to go from our side.

clalancette commented on tf2_kdl/test/test_tf2_kdl.cpp:
```diff
@@ -109,23 +107,26 @@ TEST(TfKDL, ConvertVector)
+  // auto node = rclcpp::Node::make_shared("test");
+  rclcpp::init(argc, argv);
+  rclcpp::Clock::SharedPtr clock = std::make_shared<rclcpp::Clock>(RCL_SYSTEM_TIME);
+  tf_buffer = new tf2_ros::Buffer(clock);
```
I think this would be better as a `std::unique_ptr`.  Then we would make sure not to leak memory below.

clalancette commented on tf2_kdl/test/test_tf2_kdl.cpp:
```diff
@@ -33,16 +33,16 @@
-
+#include <tf2/convert.h>
 
 tf2_ros::Buffer* tf_buffer;
```
See comment below about making this a `std::unique_ptr`.

clalancette commented on tf2_kdl/include/tf2_kdl/tf2_kdl.h:
```diff
@@ -59,16 +62,46 @@ KDL::Frame transformToKDL(const geometry_msgs::TransformStamped& t)
+ * \return The TimePoint conversion.
+ */
+inline
+tf2::TimePoint timePointFromTime(const builtin_interfaces::msg::Time& t)
```
We already have versions of this and `timeFromTimePoint` in https://github.com/ros2/geometry2/blob/ros2/tf2_ros/include/tf2_ros/buffer_interface.h#L46 ; is there any reason we can't use those?

ahcorde replied:
Done, But I was wondering if this methods should go into `tf2/time.h`. inside `tf2_ros/buffer_interface.h` it's not very intuitive

clalancette replied:
> Done, But I was wondering if this methods should go into `tf2/time.h`. inside `tf2_ros/buffer_interface.h` it's not very intuitive

The reason for that is that `tf2` itself is meant to be ROS-agnostic; you'll notice that there are no includes of ROS header files anywhere directly in tf2.  The interface between tf2 and ROS comes via the tf2_ros package.

That being said, I agree that it is not very intuitive that the time functions are in `tf2_ros/buffer_interface.h`.  I'd be in favor of making a new file `tf2_ros/time.h` where these methods would go, but I'll suggest we do that in a separate PR.

ahcorde replied:
Sure! I like `tf2_ros/time.h`

clalancette reviewed (changes requested):
There are a few things to fixup here, but nothing major.

ahcorde committed:
Using timeFromTimePoint function from tf2_ros2

ahcorde committed:
Fixing tf2_kdl test memory leaks

clalancette commented on tf2_kdl/test/test_tf2_kdl.cpp:
```diff
@@ -109,23 +107,23 @@ TEST(TfKDL, ConvertVector)
 
   int retval = RUN_ALL_TESTS();
-  delete tf_buffer;
+  tf_buffer.reset();
```
With the change to a `unique_ptr`, this shouldn't be necessary; it will be destroyed as it goes out of scope.  So you can just remove this.

clalancette replied:
This is still unresolved in the latest patches.

ahcorde replied:
Done

clalancette reviewed (changes requested):
One more minor change, then this is good to go from me.  I'll run CI on it in the meantime.

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6825)](http://ci.ros2.org/job/ci_linux/6825/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3121)](http://ci.ros2.org/job/ci_linux-aarch64/3121/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5613)](http://ci.ros2.org/job/ci_osx/5613/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6651)](http://ci.ros2.org/job/ci_windows/6651/)

vmayoral commented:
@clalancette could you please consider https://github.com/ros2/geometry2/pull/93 as well (required for moveit2 as well)

clalancette commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -34,19 +34,33 @@
+ */
+ inline
+ Eigen::Isometry3d transformToEigen(const geometry_msgs::msg::Transform& t) {
+ return Eigen::Isometry3d(Eigen::Translation3d(t.translation.x, t.translation.y, t.translation.z)
```
One other nit: need two spaces before the `return` here.

ahcorde committed:
indent return

clalancette commented:
It looks like Windows is upset for two different reasons:

```
15:19:18 C:\J\workspace\ci_windows\ws\src\ros2\geometry2\tf2_kdl\test\test_tf2_kdl.cpp(45): error C2065: 'M_PI': undeclared identifier [C:\J\workspace\ci_windows\ws\build\tf2_kdl\test_kdl.vcxproj]
15:19:18 C:\J\workspace\ci_windows\ws\src\ros2\geometry2\tf2_kdl\test\test_tf2_kdl.cpp(120): warning C4244: '=': conversion from 'double' to 'builtin_interfaces::msg::Time_<std::allocator<void>>::_sec_type', possible loss of data [C:\J\workspace\ci_windows\ws\build\tf2_kdl\test_kdl.vcxproj]
15:19:18 C:\J\workspace\ci_windows\ws\src\ros2\geometry2\tf2_kdl\test\test_tf2_kdl.cpp(121): warning C4244: '=': conversion from 'double' to 'builtin_interfaces::msg::Time_<std::allocator<void>>::_nanosec_type', possible loss of data [C:\J\workspace\ci_windows\ws\build\tf2_kdl\test_kdl.vcxproj]
```

The M_PI thing we've dealt with before, but I can't exactly remember how; @tfoote , do you remember?  The loss of precision thing is actually true, so we either need to explicitly cast it or do something smarter there.

tfoote commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -186,10 +314,58 @@ geometry_msgs::msg::Pose toMsg(const Eigen::Affine3d& in) {
+  msg.orientation.y = q.y();
+  msg.orientation.z = q.z();
+  msg.orientation.w = q.w();
+  if (msg.orientation.w < 0) {
```
Renormalizing the quaternion here will break the the one to one mapping I'd expect from a conversions. A -> B -> A should return the same thing.

tfoote commented on tf2_kdl/test/test_tf2_kdl.cpp:
```diff
@@ -33,16 +33,16 @@
 {
-  tf2::Stamped<KDL::Frame> v1(KDL::Frame(KDL::Rotation::RPY(M_PI, 0, 0), KDL::Vector(1,2,3)), builtin_interfaces::msg::Time(2.0), "A");
-
+  tf2::Stamped<KDL::Frame> v1(KDL::Frame(KDL::Rotation::RPY(M_PI, 0, 0), KDL::Vector(1,2,3)), tf2::TimePoint(tf2::durationFromSec(2.0)), "A");
```
The windows M_PI warning can be worked around with an extra #define 

See https://github.com/ros2/geometry2/commit/dc7b000 for an example

clalancette replied:
This should be resolved now (I'll run CI again to ensure it).

tfoote commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -186,10 +314,58 @@ geometry_msgs::msg::Pose toMsg(const Eigen::Affine3d& in) {
+  msg.orientation.y = q.y();
+  msg.orientation.z = q.z();
+  msg.orientation.w = q.w();
+  if (msg.orientation.w < 0) {
```
Same for renormalization above. The original form should be preserved.

ahcorde committed:
Define _USE_MATH_DEFINES so Windows gets M_PI symbol.

ahcorde committed:
Quaternion one to one mapping

clalancette commented:
Actually, before I run CI again, there hasn't been any change for the Windows loss of precision, so @ahcorde will need to do something about that before moving forward.

ahcorde committed:
Removing delete of a unique_ptr object

ahcorde committed:
Fixing possible loss of data in windows

ahcorde commented:
can you try the CI with this last commit? 

`header.stamp` contains:

```
int32 sec
uint32 nanosec
```

I was assigning a double. Probably the sizeof(double) and the sizeof(int32) (or sizeof(uint32)) is different and that's why it's failing.

clalancette commented:
Rebuild of just Windows: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6691)](http://ci.ros2.org/job/ci_windows/6691/)

If this succeeds, I'll do one more CI run of everything.

clalancette commented:
All right, Windows is happy, here is the whole lot:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6871)](http://ci.ros2.org/job/ci_linux/6871/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3157)](http://ci.ros2.org/job/ci_linux-aarch64/3157/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5644)](http://ci.ros2.org/job/ci_osx/5644/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6691)](http://ci.ros2.org/job/ci_windows/6691/)

clalancette commented on tf2_kdl/include/tf2_kdl/tf2_kdl.h:
```diff
@@ -33,12 +33,16 @@
+#include <builtin_interfaces/msg/time.hpp>
+#include <geometry_msgs/msg/transform_stamped.hpp>
 
+static const float NSEC_PER_SECOND = 1e+9;
```
We no longer need this now that we are using the other functions.

clalancette reviewed (commented):
One last nitpick, then I'm happy to approve.

clalancette committed:
Remove unused NSEC_PER_SECOND.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
All right, I removed the NSEC_PER_SECOND in 21a8551, and did a local build (the change is simple enough that I don't think CI is warranted here).  I'm going to approve, @tfoote I'll leave it to you to approve and merge.

tfoote commented:
Overall we're very close. The diff looks good. However I was just reading through the history and noticed that several items flagged by @mlautman as good things todo from earlier in this thread that have not been addressed: https://github.com/ros2/geometry2/pull/90#issuecomment-474060368

I checked off orocos_kdl as it's there. But the other ones should also be addressed, especially the rospy usage and deprecated header. The test porting would be great to do but that's improving beyond the existing functionality.

ahcorde committed:
Adding ament_python_install_package to tf2_kdl

ahcorde committed:
Remove deprecated tf2_kdl/include/tf2_kdl/tf2_kdl subdirectory

ahcorde commented:
I have been trying to execute tf2_kdl tests in python. But right now there is no a official port in the main branch of `tf2_ros` and `tf2_py` or the action in `tf2_msgs`. I have been checking others PR like this: https://github.com/ros2/geometry2/pull/99. it seems that these libraries are working properly. 

But adding these libraries I'm still stuck on this line 

```
out = b.transform(tf2_ros.Stamped(v, builtin_interfaces.msg.Time(sec=2), 'a'), 'b')
```

because of this error:

```
Traceback (most recent call last):
  File "/home/ahcorde/moveit_ws/src/geometry2/tf2_kdl/scripts/test.py", line 29, in test_transform
    out = b.transform(tf2_ros.Stamped(v, builtin_interfaces.msg.Time(sec=2), 'a'), 'b')
  File "/home/ahcorde/moveit_ws/install/lib/python3.6/site-packages/tf2_ros/buffer_interface.py", line 64, in transform
    do_transform = self.registration.get(type(object_stamped))
  File "/home/ahcorde/moveit_ws/install/lib/python3.6/site-packages/tf2_ros/buffer_interface.py", line 203, in get
    raise TypeException('Type %s if not loaded or supported'% str(key))
tf2_ros.buffer_interface.TypeException: Type <class 'PyKDL.Vector'> if not loaded or supported
```

Any ideas?

ahcorde committed:
Porting tf2_kdl python test

ahcorde committed:
Porting tf2_kdl pythonn library

ahcorde commented:
I was able to fix the tf2_kdl test in python. But as I mentioned before I had to cherry pick some of the commits in this PR https://github.com/ros2/geometry2/pull/99.

Is there any chance to revisit this PR https://github.com/ros2/geometry2/pull/99 and merge it? otherwise this Python test is not going to work

clalancette committed:
Cleanup the package.xml and CMakeLists.txt

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
> Is there any chance to revisit this PR #99 and merge it? otherwise this Python test is not going to work

Sorry for being quiet here.  After a bit of discussion with @tfoote , I think we've determined that for now it will be fine to just not build and install the python stuff, and leave that as a TODO for later.  So we can forget about #99 in terms of this PR; I'll open an issue on this repository to finish that porting.

Towards getting things finished, I've taken the liberty of rebasing this PR and doing some cleanup in the CMakeLists.txt and package.xml to get things moving again.  Take a look at what I've done in ec63e8b, and let me know what you think.

I'm also going to run CI on this shortly.

clalancette commented:
@vmayoral Apologies, we had a collision with 3bbfee556d5b8015d4c5f704005a9db5cb8b8d5e and my rebase.  I'll restore your changes in a few minutes.

clalancette commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7035)](http://ci.ros2.org/job/ci_linux/7035/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3270)](http://ci.ros2.org/job/ci_linux-aarch64/3270/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5765)](http://ci.ros2.org/job/ci_osx/5765/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6854)](http://ci.ros2.org/job/ci_windows/6854/)

clalancette commented:
And issue for python port opened in https://github.com/ros2/geometry2/issues/110

clalancette commented:
All right, CI is all green and things seem happy.  I'm going to give this 24 hours for any further objections, then I'll merge it.

ahcorde commented:
@clalancette Thanks for reviewing it :+1:

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/the-moveit-2-journey-part-1-porting-and-understanding-moveit-core/8718/1
